ambry-frontend/src/main/java/com.github.ambry.frontend/PostBlobHandler.java

```diff
@@ -0,0 +1,212 @@
+   * that will be returned in the "Location" header.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param restResponseChannel the {@link RestResponseChannel} where headers should be set.
+   * @param blobInfo the {@link BlobInfo} to make the router call with.
+   * @param finalCallback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @return a {@link Callback} to be used with {@link Router#putBlob}.
+   */
+  private Callback<String> routerPutBlobCallback(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      BlobInfo blobInfo, Callback<ReadableStreamChannel> finalCallback) {
+    CallbackTracker callbackTracker =
+        new CallbackTracker(restRequest, LOGGER, frontendMetrics.postRouterPutBlobMetrics);
+    ThrowingConsumer<String> successAction = blobId -> idConverter.convert(restRequest, blobId,
+        idConverterCallback(restRequest, restResponseChannel, blobInfo, finalCallback));
+    return FrontendUtils.managedCallback(callbackTracker, finalCallback, successAction);
+  }
+
+  /**
+   * After {@link IdConverter#convert} finishes, set the "Location" header and call
+   * {@link SecurityService#processResponse}.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param restResponseChannel the {@link RestResponseChannel} where headers should be set.
+   * @param blobInfo the {@link BlobInfo} to use for security checks.
+   * @param finalCallback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @return a {@link Callback} to be used with {@link SecurityService#processResponse}.
+   */
+  private Callback<String> idConverterCallback(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      BlobInfo blobInfo, Callback<ReadableStreamChannel> finalCallback) {
+    CallbackTracker callbackTracker = new CallbackTracker(restRequest, LOGGER, frontendMetrics.postIdConversionMetrics);
+    ThrowingConsumer<String> successAction = convertedBlobId -> {
+      restResponseChannel.setHeader(RestUtils.Headers.LOCATION, convertedBlobId);
+      securityService.processResponse(restRequest, restResponseChannel, blobInfo,
+          securityProcessResponseCallback(restRequest, finalCallback));
+    };
+    return FrontendUtils.managedCallback(callbackTracker, finalCallback, successAction);
+  }
+
+  /**
+   * After {@link SecurityService#processResponse}, call the provided callback.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param finalCallback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @return a {@link Callback} to be used with {@link SecurityService#processResponse}.
+   */
+  private Callback<Void> securityProcessResponseCallback(RestRequest restRequest,
+      Callback<ReadableStreamChannel> finalCallback) {
```

Comment:
I think it is better to have the `finalCallback` as type `Callback<String>` (because that is what a POST is expected to return). The callback impl probably doesn't care for the result today but if it did, it could use the blob ID.

Reply:
Let me think about this. One concern I had is that the name of the header that we want to set, "Location", could be considered business logic specific to postBlob, and I wanted to centralize that logic in the handler class.

Reply:
`finalCallback ` is passed in a callback chain. I am thinking if we can set it as class member? 
The main benefit is for code readability. To have less struggle between `successAction` and 'finalCallback'.

Reply:
Maybe `successAction` can be rename as 'nextActionIfNoException' or  'nextActionIfSuccess'.

Reply:
@zzmao Unfortunately the `finalCallback` has request specific dependencies (`AmbryBlobStorageService.submitResponse(...)` requires the `RestRequest` and `RestResponseChannel`). One thing that I could do is create a context object that includes the 3 immutable things (`restRequest`, `restResponseChannel`, and `finalCallback`) and pass it through the callback chain. Do you think that would help make things cleaner?

Reply:
Set `finalCallback ` in `PostBlobHandler` constructor is hard. But looks possible inside  `postBlobHandler.handle()`: 
'''
  void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
      Callback<ReadableStreamChannel> callback) {
    this.finalCallback = callback;
    ....
}
'''
Actually, all of (`restRequest`, `restResponseChannel`, and `finalCallback`) can be set inside `postBlobHandler.handle()`, then callbacks get them via `this.*` .

Reply:
@zzmao That won't work. There is only instance of the `PostBlobHandler` and all requests are processed by it. It cannot have any request specific class members since it is not an instance per request.

Reply:
Yes @zzmao, @vgkholla is right in that the handler can handle multiple requests simultaneously, so these fields cannot simply be a member of `PostBlobHandler`. 

One idea is to represent the callback chain as an inner class, and we would construct an instance of this inner class for every request we handle. Take a look at this [gist](https://gist.github.com/cgtz/1952dbaad1f1da08d08c5adc657b37b7). Would you prefer this to passing everything as function args?

Reply:
I got it. Thanks for explaining! 

@cgtz The gist looks good. I personally prefer this. How about your guys?